components/cli/src/password.rs

```diff
@@ -0,0 +1,235 @@
+// Copyright 2019 The Exonum Team
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! This module contains utilities for passphrase entry.
+
+use failure::{bail, Error, ResultExt};
+use rpassword::read_password_from_tty;
+use serde::{Deserialize, Serialize};
+use zeroize::Zeroize;
+
+use std::ops::{Deref, DerefMut};
+use std::{env, str::FromStr};
+
+/// Default name of the environment variable with a consensus key passphrase.
+pub const DEFAULT_CONSENSUS_PASS_ENV_VAR: &str = "EXONUM_CONSENSUS_PASS";
+/// Default name of the environment variable with a service key passphrase.
+pub const DEFAULT_SERVICE_PASS_ENV_VAR: &str = "EXONUM_SERVICE_PASS";
+
+/// A wrapper around `String` which securely erases itself on drop.
+#[derive(Debug, Default, Serialize, Deserialize, PartialEq)]
+pub struct Passphrase(pub String);
+
+impl Drop for Passphrase {
+    fn drop(&mut self) {
+        self.0.zeroize()
+    }
+}
+
+impl Deref for Passphrase {
```

Comment:
Usually deref implementation is anti-pattern in Rust, may you look carefully whether deref is need here.

Reply:
I'm not aware of `Deref` being an anti-pattern (besides the cases where it is used to implement unnecessary inheritance). Moreover, `Deref` is usually used to implement smart-pointers functionality, and `Passphrase` can be seen as a smart-pointer.

Reply:
However, I decided to rewrite it to make it more safe.